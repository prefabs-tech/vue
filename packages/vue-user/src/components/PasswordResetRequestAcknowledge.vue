<template>
  <div class="acknowledgement-content">
    <i18n-t
      class="acknowledgement-message"
      keypath="user.passwordResetAcknowledge.message"
      tag="p"
    >
      <span class="email">{{ email }}</span>
    </i18n-t>

    <div class="resend">
      <span>{{ t("user.passwordResetAcknowledge.label.notReceived") }}</span>

      <i18n-t
        v-if="timer > 0"
        class="resen-timer disabled"
        keypath="user.passwordResetAcknowledge.label.resendIn"
        tag="span"
      >
        {{ formatDuration(timer) }}
      </i18n-t>

      <span v-else class="resend-link" @click="$emit('resend')">
        {{ t("user.passwordResetAcknowledge.actions.resend") }}
      </span>
    </div>
  </div>
</template>

<script lang="ts">
export default {
  name: "PasswordResetRequestAcknowledge",
};
</script>

<script setup lang="ts">
import { useConfig } from "@prefabs.tech/vue3-config";
import { useI18n } from "@prefabs.tech/vue3-i18n";
import { formatDuration } from "@prefabs.tech/vue3-ui";
import { onMounted, ref } from "vue";

import { useTranslations } from "../index";

import type { AppConfig } from "@prefabs.tech/vue3-config";

const { user: userConfig } = useConfig() as AppConfig;

const messages = useTranslations();

const { t } = useI18n({ messages });

defineProps({
  email: {
    required: true,
    type: String,
  },
});

defineEmits(["resend"]);

const timer = ref<number>(
  userConfig?.features?.forgotPasswordResendTimeInSeconds ?? 30,
);

const redirectCountDown = () => {
  setTimeout(() => {
    timer.value--;
    redirectCountDown();
  }, 1000);
};

onMounted(() => {
  redirectCountDown();
});
</script>

<style lang="css">
.acknowledgement-content {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.acknowledgement-content .email {
  font-weight: 600;
}

.acknowledgement-content .resend {
  display: flex;
  font-style: italic;
  align-items: baseline;
  gap: 0.25rem;
}

.acknowledgement-content .resend .resend-link {
  color: var(--dz-primary-color, #007aff);
  cursor: pointer;
}

.acknowledgement-content .resend .resend-link:hover {
  text-decoration: underline;
}

.acknowledgement-content .resend .disabled {
  color: var(--color);
  cursor: not-allowed;
  opacity: 0.8;
  text-decoration: none;
}
</style>
