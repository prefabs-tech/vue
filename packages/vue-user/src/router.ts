import { storeToRefs } from "pinia";
import { Router } from "vue-router";

import AuthSocialLoginCallback from "./components/AuthSocialLoginCallback.vue";
import useUserStore from "./store";
import AcceptInvitation from "./views/AcceptInvitation.vue";
import ChangePassword from "./views/ChangePassword.vue";
import VerifyEmailReminder from "./views/EmailVerificationReminder.vue";
import Login from "./views/Login.vue";
import PasswordReset from "./views/PasswordReset.vue";
import PasswordResetRequest from "./views/PasswordResetRequest.vue";
import Profile from "./views/Profile.vue";
import Roles from "./views/Roles/Index.vue";
import Signup from "./views/Signup.vue";
import SignupFirstUser from "./views/SignupFirstUser.vue";
import VerifyEmail from "./views/VerifyEmail.vue";

import type {
  DzangolabVueUserConfig,
  RouteOverride,
  RouteOverrides,
} from "./types";
import type { RouteMeta, RouteRecordRaw } from "vue-router";

const _routes = {
  acceptInvitation: {
    component: AcceptInvitation,
    name: "acceptInvitation",
    path: "/signup/token/:token?",
  } as RouteRecordRaw,
  changePassword: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: ChangePassword,
    name: "changePassword",
    path: "/change-password",
  },
  facebook: {
    component: AuthSocialLoginCallback,
    name: "authFacebookCallback",
    path: "/auth/callback/facebook",
  },
  google: {
    component: AuthSocialLoginCallback,
    name: "authGoogleCallback",
    path: "/auth/callback/google",
  },
  login: {
    component: Login,
    name: "login",
    path: "/login",
  } as RouteRecordRaw,
  profile: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: Profile,
    name: "profile",
    path: "/profile",
  } as RouteRecordRaw,
  roles: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: Roles,
    name: "roles",
    path: "/roles",
  } as RouteRecordRaw,
  signup: {
    component: Signup,
    name: "signup",
    path: "/signup",
  } as RouteRecordRaw,
  signupFirstUser: {
    component: SignupFirstUser,
    name: "signupFirstUser",
    path: "/signup-first-user",
  } as RouteRecordRaw,
  passwordReset: {
    component: PasswordReset,
    name: "resetPassword",
    path: "/reset-password/:token?",
  } as RouteRecordRaw,
  passwordResetRequest: {
    component: PasswordResetRequest,
    name: "resetPasswordRequest",
    path: "/reset-password-request",
  } as RouteRecordRaw,
  verifyEmail: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: VerifyEmail,
    name: "verifyEmail",
    path: "/verify-email",
  },
  verifyEmailReminder: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: VerifyEmailReminder,
    name: "verifyEmailReminder",
    path: "/verify-email-reminder",
  },
};

const getRoute = (
  defaultRoute: object,
  override?: RouteOverride,
): RouteRecordRaw => {
  return {
    ...defaultRoute,
    ...override,
  } as RouteRecordRaw;
};

const addRoutes = (router: Router, userConfig?: DzangolabVueUserConfig) => {
  const routes: RouteOverrides | undefined = userConfig?.routes;

  router.addRoute(getRoute(_routes.facebook, routes?.facebook));

  router.addRoute(getRoute(_routes.google, routes?.google));

  router.addRoute(getRoute(_routes.login, routes?.login));

  if (!routes?.signup?.disabled) {
    router.addRoute(getRoute(_routes.signup, routes?.signup));
  }

  if (routes?.signup?.disabled && !routes?.signupFirstUser?.disabled) {
    router.addRoute(getRoute(_routes.signupFirstUser, routes?.signupFirstUser));
  }

  router.addRoute(getRoute(_routes.passwordReset, routes?.passwordReset));

  router.addRoute(
    getRoute(_routes.passwordResetRequest, routes?.passwordResetRequest),
  );

  router.addRoute(getRoute(_routes.profile, routes?.profile));

  router.addRoute(getRoute(_routes.roles, routes?.roles));

  router.addRoute(
    getRoute(
      _routes.changePassword,
      routes?.changePassword,
    ),
  );

  router.addRoute(getRoute(_routes.acceptInvitation, routes?.acceptInvitation));

  if (userConfig?.features?.signUp?.emailVerification) {
    router.addRoute(getRoute(_routes.verifyEmail, routes?.verifyEmail));

    router.addRoute(
      getRoute(_routes.verifyEmailReminder, routes?.verifyEmailReminder),
    );
  }
};

const redirectRoutes = (router: Router) => {
  router.beforeEach((to, from, next) => {
    const userStore = useUserStore();
    const { user } = storeToRefs(userStore);

    const routesToRedirect = [
      "acceptInvitation",
      "login",
      "signup",
      "signupFirstUser",
      "resetPassword",
      "resetPasswordRequest",
    ];
    const name = to.name as string;

    if (user.value && routesToRedirect.includes(name)) {
      next({ name: "profile" });
    } else {
      next();
    }
  });
};

const addAuthenticationGuard = (
  router: Router,
  userConfig?: DzangolabVueUserConfig,
) => {
  router.beforeEach(async (to) => {
    const meta = to.meta as RouteMeta;

    const userStore = useUserStore();

    const EmailVerificationEnabled =
      !!userConfig?.features?.signUp?.emailVerification;

    const name = to.name as string;
    const routesToRedirect = ["verifyEmail", "verifyEmailReminder"];
    const { user } = storeToRefs(userStore);
    const { getUser, getVerificationStatus } = userStore;

    if (!user.value) {
      user.value = await getUser();
    }

    const isSocialLoggedIn = user.value?.thirdParty && userConfig?.socialLogins?.includes(user.value?.thirdParty?.id);

    if (meta.authenticated && !user.value) {
      router.push({ name: "login" }); // using next inside async function is not allowed
    }

    if (isSocialLoggedIn && name === "changePassword") {
      router.push({ name: "home" });
    }

    if (meta.authenticated && EmailVerificationEnabled && user.value) {
      const isEmailVerified = await getVerificationStatus();

      if (
        !isEmailVerified &&
        ![...routesToRedirect, "profile"].includes(name)
      ) {
        router.push({ name: "verifyEmailReminder" });

        return;
      } else if (isEmailVerified && routesToRedirect[1] === name) {
        router.push({ name: "home" });
      }
    }

    const isProfileCompleted = !!user.value?.isProfileCompleted;
    const profileCompletionEnabled = user.value?.isProfileCompleted !== undefined;

    if (
      meta.authenticated &&
      profileCompletionEnabled &&
      !isProfileCompleted &&
      ![...routesToRedirect, "profile"].includes(name)
    ) {
      router.push({ name: "profile" });
    }
  });
};

const updateRouter = (router: Router, userConfig?: DzangolabVueUserConfig) => {
  addRoutes(router, userConfig);

  addAuthenticationGuard(router, userConfig);

  redirectRoutes(router);
};

export default updateRouter;
