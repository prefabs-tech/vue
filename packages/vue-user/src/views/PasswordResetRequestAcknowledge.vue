<template>
  <Page
    class="auth password-reset-acknowledge"
    :title="t('user.passwordResetAcknowledge.title')"
  >
    <i18n-t
      class="acknowledge"
      keypath="user.passwordResetAcknowledge.content"
      tag="p"
    >
      <span class="email">{{ route.query.email }}</span>
    </i18n-t>

    <nav class="links">
      <p class="resend-email">
        {{ t("user.passwordResetAcknowledge.label.notReceived") }}
        <RouterLink :to="{ name: 'resetPasswordRequest' }">
          {{ t("user.passwordResetAcknowledge.actions.resend") }}
        </RouterLink>
      </p>
    </nav>

    <i18n-t
      class="redirect"
      keypath="user.passwordResetAcknowledge.label.redirect"
      tag="p"
    >
      <span class="seconds">{{ seconds }}</span>
    </i18n-t>
  </Page>
</template>

<script lang="ts">
export default {
  name: "PasswordResetRequestAcknowledge",
};
</script>

<script setup lang="ts">
import { useI18n } from "@prefabs.tech/vue3-i18n";
import { Page } from "@prefabs.tech/vue3-ui";
import { onMounted, ref } from "vue";
import { useRoute, useRouter } from "vue-router";

import { useTranslations } from "../index";

import "../assets/css/auth.css";

const messages = useTranslations();

const { t } = useI18n({ messages });

const route = useRoute();
const router = useRouter();

let seconds = ref<number>(30);

const redirectCountDown = () => {
  if (!route.query.email) {
    seconds.value = 0;
    router.push({ name: "resetPasswordRequest" });
  }

  if (seconds.value > 0) {
    setTimeout(() => {
      seconds.value--;
      redirectCountDown();
    }, 1000);
  } else {
    router.push({ name: "resetPasswordRequest" });
  }
};

onMounted(() => {
  redirectCountDown();
});
</script>

<style lang="css">
.password-reset-acknowledge .resend-email {
  font-style: italic;
}

.password-reset-acknowledge .resend-email a {
  color: var(--dz-primary-color, #007aff);
  cursor: pointer;
  text-decoration: none;
}
</style>
