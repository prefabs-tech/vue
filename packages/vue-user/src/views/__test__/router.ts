import { createRouter, createWebHistory } from "vue-router";

import Login from "../Login.vue";

const Test = {
  name: "Test",
  template: `<h1>Test</h1>`,
};

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      component: Login,
      name: "login",
      path: "/login",
    },
    {
      component: Test,
      name: "resetPasswordRequest",
      path: "/reset-password-request",
    },
    {
      component: Test,
      path: "/",
    },
  ],
});

export default router;
