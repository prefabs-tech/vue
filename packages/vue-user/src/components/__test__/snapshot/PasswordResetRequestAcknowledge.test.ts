import configPlugin from "@prefabs.tech/vue3-config";
import i18Plugin, { useLocaleStore } from "@prefabs.tech/vue3-i18n";
import { LoadingButton, LoadingIcon } from "@prefabs.tech/vue3-ui";
import { mount, RouterLinkStub } from "@vue/test-utils";
import { createPinia } from "pinia";
import { describe, it, expect } from "vitest";

import appConfig from "../config";
import PasswordResetRequestAcknowledge from "../../PasswordResetRequestAcknowledge.vue";

import router from "../../../views/__test__/router";

import type { VueWrapper } from "@vue/test-utils";

describe("PasswordResetRequestAcknowledge", () => {
  const pinia = createPinia();

  const { setLocale } = useLocaleStore(appConfig.slug);

  const locales = ["en", "fr"];

  for (const locale of locales) {
    setLocale(locale);

    const wrapper: VueWrapper = mount(PasswordResetRequestAcknowledge, {
      global: {
        components: {
          LoadingButton,
          LoadingIcon,
        },
        plugins: [
          [
            configPlugin,
            {
              config: appConfig,
            },
          ],
          pinia,
          [
            i18Plugin,
            {
              config: appConfig,
            },
          ],
          router,
        ],
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    it("matches snapshot in " + locale, () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  }
});
