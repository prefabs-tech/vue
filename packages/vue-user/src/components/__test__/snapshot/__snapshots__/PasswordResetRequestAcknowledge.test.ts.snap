// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PasswordResetRequestAcknowledge > matches snapshot in en 1`] = `
<div
  class="acknowledgement-content"
>
  <p
    class="acknowledgement-message"
  >
    An email with a link to reset your password was sent to 
    <span
      class="email"
    />
    . Follow the instructions in the email to access your account.
  </p>
  <div
    class="resend"
  >
    <span>
      Didn't receive the email? 
    </span>
    <span
      class="resend-link"
    >
      Resend
    </span>
  </div>
</div>
`;

exports[`PasswordResetRequestAcknowledge > matches snapshot in fr 1`] = `
<div
  class="acknowledgement-content"
>
  <p
    class="acknowledgement-message"
  >
    An email with a link to reset your password was sent to 
    <span
      class="email"
    />
    . Follow the instructions in the email to access your account.
  </p>
  <div
    class="resend"
  >
    <span>
      Didn't receive the email? 
    </span>
    <span
      class="resend-link"
    >
      Resend
    </span>
  </div>
</div>
`;
